### PR TITLE
Bug 1218374 - Properly stop the conversation rendering when entering …

### PR DESCRIPTION
**Pull request by julienw**
…another conversation r=azasypkin

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bd127661b347f5a93655f22dc8fa3b129c7a1412

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=acc30e657ff3c1f64c4fcd32a739ef18f66359a2

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4725e3c8e0372e36213c8b9a963ab61d78a6cb86

File: apps/sms/test/marionette/generators/thread.js [hunk 1]
```diff
@@ -59,7 +59,7 @@ var ThreadGenerator = {
 
     for (var i = 0; i < parameters.numberOfMessages; i++) {
       var message = null;
-      var messageBody = thread.body;
+      var messageBody = thread.body + ' (threadId: ' + thread.id + ')';
```
Comment (julienw) [inline apps/sms/test/marionette/generators/thread.js]: this makes it easier to see if the message is in the right thread id when debugging

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fe163080c42b7de902d2ecab1a131bca64cb9351

File: apps/sms/test/marionette/lib/views/conversation/message_accessors.js [hunk 2]
```diff
@@ -34,7 +34,7 @@ MessageAccessor.prototype = {
       );
 
       return {
-        id: messageNode.getAttribute('data-message-id'),
+        id: +messageNode.getAttribute('data-message-id'),
```
Comment (julienw) [inline apps/sms/test/marionette/lib/views/conversation/message_accessors.js]: the id is a number :)

File: apps/sms/test/marionette/notification_test.js [hunk 3]
```diff
@@ -28,6 +29,29 @@ marionette('Message notification tests', function() {
...
+    assert.equal(
+      messagesApp.Conversation.headerTitle.text(), smsMessage.sender
+    );
+
+    // When we open thread from notification, message input should not be
+    // focused.
+    assertIsNotFocused(
+      messagesApp.Composer.messageInput, 'Message input should not be focused'
+    );
+  }
```
Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: I changed this in a previous version of the patch, and then I left it even than I don't need it. Tell me if you'd rather want that I revert the change.

File: apps/sms/test/marionette/notification_test.js [hunk 4]
```diff
@@ -225,7 +226,63 @@ marionette('Message notification tests', function() {
...
+
+    test('Clicking a notification while a conversation is loading', function() {
+      messagesApp.launch();
+      storage.setMessagesStorage(messagesStorage);
+      // Set empty contacts store.
+      storage.setContactsStorage();
+
+      var inbox = new InboxView(client);
+      var conversation = inbox.goToConversation(
+        thread1.id, { shouldWait: false }
```
Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: we need to go as fast as possible to trigger the bug

Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: without this change, the test is not failing on master :/

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4c6fc224bfabbe652c78c901f8506ae5d2728568

File: apps/sms/views/conversation/js/conversation.js [hunk 5]
```diff
@@ -1802,9 +1813,9 @@ var ConversationView = {
     if (lateness < limits.IN_HOURS)       { unit = MINUTE; } 
     else if (lateness < limits.IN_DAYS)   { unit = HOUR; } 
     else if (lateness < limits.IN_MONTHS) { unit = DAY; }
-    
+
     lateness = Math.floor(lateness / unit);
-    
+
```
Comment (julienw) [inline apps/sms/views/conversation/js/conversation.js]: some white space changes from an earlier bug

File: apps/sms/views/conversation/test/unit/conversation_test.js [hunk 6]
```diff
@@ -160,6 +160,21 @@ suite('conversation.js >', function() {
...
+    Threads.get.withArgs(threadId).returns(mockThread);
+
+    var transitionArgs = {
+      id: strThreadId,
+      meta: { next: { panel: 'thread', id: strThreadId }}
+    };
+    return ConversationView.beforeEnter(transitionArgs).then(
+      () => ConversationView.afterEnter(transitionArgs)
+    );
+  }
```
Comment (julienw) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: quite complex setup to have a proper state

File: apps/sms/views/conversation/js/conversation.js [hunk 7]
```diff
@@ -1468,6 +1468,17 @@ var ConversationView = {
     this._stopRenderingNextStep = true;
   },
 
+  /**
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __nit (:bike: ):__ It seems would be better to have method description as well (more in description and less for return statement description)

File: apps/sms/views/conversation/js/conversation.js [hunk 8]
```diff
@@ -1519,13 +1530,13 @@ var ConversationView = {
...
     var onRenderMessage = (function renderMessage(message) {
-      if (this._stopRenderingNextStep) {
+      if (!this.shouldRenderForThreadId(threadId)) {
         // stop the iteration and clear the taskQueue
         taskQueue = null;
         return false;
       }
       taskQueue.push(() => {
-        if (!this._stopRenderingNextStep) {
+        if (this.shouldRenderForThreadId(threadId)) {
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: maybe someday our TaskQueue can have ```clear/cancel``` method that will automagically not allow to execute next step and return rejected promise, however there likely be overhead for the wrapper function that will perform this check :)

File: apps/sms/test/marionette/generators/thread.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/test/marionette/generators/thread.js]: yeah, makes sense to have smth like this

File: apps/sms/test/marionette/lib/views/conversation/message_accessors.js (hunk 2 above)
Comment (azasypkin) [inline apps/sms/test/marionette/lib/views/conversation/message_accessors.js]: ah, nice catch!

File: apps/sms/test/marionette/lib/views/inbox/view.js [hunk 9]
```diff
@@ -49,17 +58,17 @@ InboxView.prototype = {
     return this._createNewMessageView();
   },
 
-  _createNewMessageView: function() {
+  _createNewMessageView: function(shouldWait) {
     var NewMessageView = require('../new-message/view');
     var newMessageView = new NewMessageView(this.client);
-    newMessageView.accessors.waitToAppear();
+    shouldWait && newMessageView.accessors.waitToAppear();
```
Comment (azasypkin) [inline apps/sms/test/marionette/lib/views/inbox/view.js]: __question:__ mm what was the issue with always waiting?

File: apps/sms/test/marionette/notification_test.js [hunk 10]
```diff
@@ -225,7 +226,63 @@ marionette('Message notification tests', function() {
...
+      });
+
+      messagesStorage = [ thread1, thread2 ];
+    });
+
+    test('Clicking a notification while a conversation is loading', function() {
+      messagesApp.launch();
+      storage.setMessagesStorage(messagesStorage);
+      // Set empty contacts store.
+      storage.setContactsStorage();
```
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: __nit:__ do we really need this? It seems we don't inject contacts mock here

File: apps/sms/test/marionette/notification_test.js (hunk 4 above)
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: __question:__ mmmmm, I believe there is a chance we won't have the issue on such _small_ conversations in real life as well :) What about 200-300 messages in the first thread instead? Closer to real use case and I believe it should fail :)

File: apps/sms/test/marionette/notification_test.js [hunk 11]
```diff
@@ -225,7 +226,63 @@ marionette('Message notification tests', function() {
...
+        thread1.id, { shouldWait: false }
+      );
+
+      // Receive 'notification' system message and fill up the storage.
+      messagesApp.sendSystemMessage('notification', {
+        clicked: true,
+        data: { threadId: thread2.id, id: thread2.messages[0].id }
+      });
+
+      conversation.accessors.waitToAppear();
```
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: __nit:__ looks like this one is also not needed or I'm missing something?

File: apps/sms/test/marionette/notification_test.js [hunk 12]
```diff
@@ -225,7 +226,63 @@ marionette('Message notification tests', function() {
...
+      var thread2Ids = thread2.messages.map(
+        function(message) { return message.id; }
+      );
+
+      // Our testing strategy is:
+      // 1. Wait that all messages are rendered for thread2.
+      thread2Ids.forEach(function(messageId) {
```
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: Looks like ```findMessage``` doesn't guarantee that message is rendered already so it can return null, how about new property on Conversation view instead?
```
client.waitFor(function() {
  return conversation.numberOfMessages >= thread2.messages.length;
});

// or much more slower option, but that exists already
client.waitFor(function() {
  return conversation.messages.length >= thread2.messages.length;
});
```

File: apps/sms/test/marionette/notification_test.js (hunk 4 above)
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: And maybe much less messages for the thread#2

File: apps/sms/test/marionette/notification_test.js (hunk 3 above)
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: Less unnecessary changes - less chances to break something ;) But in this case up to you, I don't mind really

Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: you're right, I'll revert this

File: apps/sms/test/marionette/notification_test.js (hunk 11 above)
Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: if I wait neither earlier nor here, I'll have the infamous "crash but can't stackwalk" message :/ Which doesn't help much. From my tests, it seems that we need at least that the window is unhidden (but we don't really care that it's been transformed)

File: apps/sms/test/marionette/notification_test.js (hunk 12 above)
Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: are you sure of this ? because it seemed it was waiting for me :/

File: apps/sms/views/conversation/js/conversation.js (hunk 8 above)
Comment (julienw) [inline apps/sms/views/conversation/js/conversation.js]: There has been lengthy debates about cancelable promises also ;)

File: apps/sms/test/marionette/notification_test.js (hunk 12 above)
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: > are you sure of this ? because it seemed it was waiting for me :/ 

Nope, but let me check :)

File: apps/sms/test/marionette/notification_test.js (hunk 4 above)
Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: ok, I tried with much more messages in the first thread, and you're right it reproduces the problem now. I'll revert the "shouldWait" changes.

File: apps/sms/test/marionette/notification_test.js (hunk 12 above)
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: Ah yeah, it waits for the default timeout and fails if it can't find, likely was thinking about ```findElements```, sorry, please ignore ```Looks like findMessage doesn't guarantee that message is rendered already so it can return null```, but I can't recall why I added null check there :/

Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: I thought it was throwing if it couldn't find the element after some timeout.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4fe3736fdb9c3c00191a57915341a84bde3eb179

File: apps/sms/test/marionette/notification_test.js [hunk 13]
```diff
@@ -228,4 +229,57 @@ marionette('Message notification tests', function() {
...
+        conversation.findMessage(messageId);
+      });
+
+      // 2. Look at all rendered messages and see if one of these is part of
+      // thread1.
+      conversation.messages(
```
Comment (azasypkin) [inline apps/sms/test/marionette/notification_test.js]: __optional nit:__ we can probable fail earlier with additional assert, but it's veeeery optional :)
```
var messages = conversation.messages({ first: thread2Ids.length + 1 });

assert.equal(messages.length, thread2Ids.length);

messages.forEach(.....)
```

Comment (julienw) [inline apps/sms/test/marionette/notification_test.js]: it's basically the same IMO :)

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=938a85afbe576f9a2af36679343701e75e459f27
